config/base/derive/src/lib.rs

```diff
@@ -0,0 +1,685 @@
+        match cursor.token_tree() {
+            Some((TokenTree::Punct(punct), next)) if punct.as_char() == ',' => Ok(next),
+            Some((other, next)) => Err(syn::Error::new(other.span(), "expected a comma")),
+            None => Ok(cursor),
+        }
+    }
+
+    fn expect_comma_or_eq_with_lit_str(
+        cursor: syn::buffer::Cursor,
+    ) -> syn::Result<(Option<syn::LitStr>, syn::buffer::Cursor)> {
+        const EXPECTED_STR_LIT: &str = r#"expected a string literal, e.g. "...""#;
+
+        let next = match cursor.token_tree() {
+            Some((TokenTree::Punct(punct), next)) if punct.as_char() == '=' => next,
+            Some((TokenTree::Punct(punct), next)) if punct.as_char() == ',' => {
+                return Ok((None, next))
+            }
+            None => return Ok((None, cursor)),
+            Some((other, _)) => Err(syn::Error::new(other.span(), "expected ',' or '='"))?,
+        };
+
+        let (lit, next) = match next.token_tree() {
+            Some((TokenTree::Literal(lit), next)) => (lit, next),
+            Some((other, _)) => Err(syn::Error::new(other.span(), EXPECTED_STR_LIT))?,
+            None => Err(syn::Error::new(next.span(), EXPECTED_STR_LIT))?,
+        };
+
+        let string = lit.to_string();
+        let trimmed = string.trim_matches('"');
+        if string == trimmed {
+            // not a string literal
+            Err(syn::Error::new(lit.span(), EXPECTED_STR_LIT))?;
+        }
+
+        let lit_str = syn::LitStr::new(trimmed, lit.span());
+        Ok((Some(lit_str), next))
+    }
+
+    #[derive(Debug, PartialEq)]
+    struct ParameterTypeShape {
+        option: bool,
+        with_origin: bool,
+    }
+
+    impl ParameterTypeShape {
+        fn analyze(ty: &syn::Type) -> Self {
+            #[derive(Debug)]
+            enum Token {
+                Option,
+                WithOrigin,
+                Unknown,
+            }
+
+            fn parse_tokens(ty: &syn::Type, depth: u8) -> Vec<Token> {
+                if depth == 0 {
+                    return vec![];
+                }
+
+                let mut found = None;
+
+                if let syn::Type::Path(type_path) = ty {
+                    if let Some(last_segment) = type_path.path.segments.last() {
+                        match &last_segment.arguments {
+                            syn::PathArguments::AngleBracketed(args) if args.args.len() == 1 => {
+                                if let syn::GenericArgument::Type(ty) =
+                                    args.args.first().expect("should be exactly 1")
+                                {
+                                    found = Some((Some(ty), &last_segment.ident));
+                                }
+                            }
+                            syn::PathArguments::None => found = Some((None, &last_segment.ident)),
+                            _ => {}
+                        }
+                    }
+                }
```

Comment:
You can return value from the match instead of assigning it to the mutable variable

```suggestion
                let found = if let syn::Type::Path(type_path) = ty {
                    if let Some(last_segment) = type_path.path.segments.last() {
                        match &last_segment.arguments {
                            syn::PathArguments::AngleBracketed(args) if args.args.len() == 1 => {
                                if let syn::GenericArgument::Type(ty) =
                                    args.args.first().expect("should be exactly 1")
                                {
                                    Some((Some(ty), &last_segment.ident))
                                }
                            }
                            syn::PathArguments::None => Some((None, &last_segment.ident)),
                            _ => None,
                        }
                    }
                }
```